File: tests/leetcode/x/go/0632.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
)

type item struct{ val, row, idx int }
type minHeap []item

func (h minHeap) Len() int            { return len(h) }
func (h minHeap) Less(i, j int) bool  { return h[i].val < h[j].val }
func (h minHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(x any)         { *h = append(*h, x.(item)) }
func (h *minHeap) Pop() any           { old := *h; x := old[len(old)-1]; *h = old[:len(old)-1]; return x }

func smallestRange(nums [][]int) (int, int) {
	h := &minHeap{}
	heap.Init(h)
	currentMax := -1 << 60
	for i, row := range nums {
		v := row[0]
		heap.Push(h, item{v, i, 0})
		if v > currentMax {
			currentMax = v
		}
	}
	bestL, bestR := (*h)[0].val, currentMax
	for {
		cur := heap.Pop(h).(item)
		if currentMax-cur.val < bestR-bestL || currentMax-cur.val == bestR-bestL && cur.val < bestL {
			bestL, bestR = cur.val, currentMax
		}
		if cur.idx+1 == len(nums[cur.row]) {
			return bestL, bestR
		}
		nv := nums[cur.row][cur.idx+1]
		heap.Push(h, item{nv, cur.row, cur.idx + 1})
		if nv > currentMax {
			currentMax = nv
		}
	}
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	var t int
	if _, err := fmt.Fscan(in, &t); err != nil {
		return
	}
	for tc := 0; tc < t; tc++ {
		var k int
		fmt.Fscan(in, &k)
		nums := make([][]int, k)
		for i := 0; i < k; i++ {
			var n int
			fmt.Fscan(in, &n)
			nums[i] = make([]int, n)
			for j := 0; j < n; j++ {
				fmt.Fscan(in, &nums[i][j])
			}
		}
		l, r := smallestRange(nums)
		if tc > 0 {
			fmt.Fprintln(out)
			fmt.Fprintln(out)
		}
		fmt.Fprintf(out, "[%d,%d]", l, r)
	}
}

File: tests/leetcode/x/cpp/0632.cpp
#include <iostream>
#include <queue>
#include <tuple>
#include <vector>
using namespace std;
pair<int,int> smallestRange(const vector<vector<int>>& nums){
    using T=tuple<int,int,int>;
    priority_queue<T,vector<T>,greater<T>> pq;
    int currentMax=-1000000007;
    for(int i=0;i<(int)nums.size();i++){pq.push({nums[i][0],i,0});currentMax=max(currentMax,nums[i][0]);}
    int bestL=get<0>(pq.top()),bestR=currentMax;
    while(true){
        auto [cur,row,idx]=pq.top();pq.pop();
        if(currentMax-cur<bestR-bestL||(currentMax-cur==bestR-bestL&&cur<bestL)){bestL=cur;bestR=currentMax;}
        if(idx+1==(int)nums[row].size()) return {bestL,bestR};
        int nv=nums[row][idx+1];
        pq.push({nv,row,idx+1});
        currentMax=max(currentMax,nv);
    }
}
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int t; if(!(cin>>t)) return 0;
    for(int tc=0;tc<t;tc++){
        int k; cin>>k;
        vector<vector<int>> nums(k);
        for(int i=0;i<k;i++){
            int n; cin>>n;
            nums[i].resize(n);
            for(int j=0;j<n;j++) cin>>nums[i][j];
        }
        auto [l,r]=smallestRange(nums);
        if(tc) cout<<"\n\n";
        cout<<"["<<l<<","<<r<<"]";
    }
}
